services/api/src/api/routes/endpoint.py

```diff
@@ -28,21 +30,124 @@
+    split: Optional[str]
+    input_type: InputType = field(init=False)
+
+    def __post_init__(self) -> None:
+        self.input_type = "split" if self.split else "config" if self.config else "dataset"
+
+
+def get_input_parameters(query_params: QueryParams) -> InputParameters:
+    dataset = query_params.get("dataset")
+
+    if not are_valid_parameters([dataset]) or not dataset:
+        raise MissingRequiredParameterError("Parameter 'dataset' is required")
+
+    config = None
+    config_param = query_params.get("config")
+    if are_valid_parameters([config_param]):
+        config = config_param
+
+    split = None
+    split_param = query_params.get("split")
+    if are_valid_parameters([split_param]):
+        split = split_param
+        if config is None:
+            raise MissingRequiredParameterError("Parameter 'config' is required")
+
+    return InputParameters(dataset=dataset, config=config, split=split)
+
+
+def get_first_succeeded_cache_entry_from_steps(
     processing_steps: List[ProcessingStep],
+    input_parameters: InputParameters,
+    init_processing_steps: List[ProcessingStep],
+    hf_endpoint: str,
+    hf_token: Optional[str] = None,
+) -> Optional[CacheEntry]:
+    """Gets the cache from the first succedeed step in the processing step list.
+    If no succeeded result is found, it will return the last one even if it was failed,
+    if no one result is found, it returns None.
+    Checks if job is still in progress by each processing step in case of no entry found.
+    """
+    last_result = None
+    for processing_step in processing_steps:
+        try:
+            last_result = get_response(
+                kind=processing_step.cache_kind,
+                dataset=input_parameters.dataset,
+                config=input_parameters.config,
+                split=input_parameters.split,
+            )
+
+            if last_result["http_status"] == HTTPStatus.OK:
+                return last_result
+        except DoesNotExist:
+            logging.warning(
+                f"processing_step={processing_step.name} dataset={input_parameters.dataset} "
+                f"config={input_parameters.config} split={input_parameters.split} no entry found"
+            )
+            try:
+                check_in_process(
+                    processing_step=processing_step,
+                    init_processing_steps=init_processing_steps,
+                    dataset=input_parameters.dataset,
+                    config=input_parameters.config,
+                    split=input_parameters.split,
+                    hf_endpoint=hf_endpoint,
+                    hf_token=hf_token,
+                )
+            except (PreviousStepError, DatasetError):
+                raise ResponseNotFoundError("Not found.")
+    return last_result
+
+
+def get_response_from_cache_entry(
+    result: Optional[CacheEntry],
+    max_age_long: int = 0,
+    max_age_short: int = 0,
+) -> Response:
+    if result:
```

Comment:
detail: I use to return as soon as possible when there is an error, to reduce the number of indented lines in the common case, that generally contains more code:

```python
if not result:
  raise ResponseNotReadyError(
        "The server is busier than usual and the response is not ready yet. Please retry later."
    )
content = result["content"]
...
```

Reply:
Done, thanks!